asf_smart_meter_exploration/pipeline/data_aggregation.py

```diff
@@ -0,0 +1,177 @@
+"""
+Functions to process household smart meter data into various formats for clustering.
+"""
+
+import numpy as np
+import holidays
+
+from asf_smart_meter_exploration.getters.get_processed_data import (
+    get_meter_data,
+    get_household_data,
+)
+
+
+meter_data = get_meter_data()
+
+
+def get_average_usage(meter_data=meter_data, normalised=False, cumulative=False):
+    """For each household, get average usage for each half-hour of the day.
+
+    Args:
+        meter_data (pd.DataFrame, optional): _description_. Defaults to meter_data.
+        normalised (bool, optional): Whether to normalise the readings and return
+            the proportion used in each half-hour rather than the amount in kWh.
+            Defaults to False.
+        cumulative (bool, optional): Whether to return the cumulative sums of the values instead.
+            Defaults to False.
+
+    Returns:
+        pd.DataFrame: Average usage data.
+    """
+
+    hh_averages = meter_data.groupby("time").mean(numeric_only=True).dropna(axis=1).T
+
+    if normalised:
+        hh_averages = hh_averages.div(hh_averages.sum(axis=1), axis=0).dropna(axis=0)
+    if cumulative:
+        hh_averages = hh_averages.cumsum(axis=1)
+
+    return hh_averages
+
+
+def get_average_usage_daytypes(meter_data=meter_data, normalise=False):
+    """For each household, get average usage split by "day type" (weekday or weekend).
+
+    Args:
+        meter_data (_type_, optional): _description_. Defaults to meter_data.
+        normalised (bool, optional): Whether to normalise the readings and return
+            the proportion used in each half-hour rather than the amount in kWh.
+            Values are normalised within each day type. Defaults to False.
+
+    Returns:
+        pd.DataFrame: Average usage data split by day type.
+    """
+
+    meter_data["day_of_week"] = meter_data["tstp"].dt.day_of_week
+    meter_data["day_type"] = np.where(
+        meter_data["day_of_week"] <= 4, "weekday", "weekend"
+    )
+
+    bank_holidays = holidays.country_holidays(
+        "UK", subdiv="England", years=[2011, 2012, 2013, 2014]
+    ).keys()
+    meter_data["bank_holiday"] = np.where(
+        meter_data["tstp"].dt.date.isin(bank_holidays), True, False
+    )
+
+    meter_data["weekend_or_bank_holiday"] = (
+        meter_data["day_type"] == "weekend"
+    ) | meter_data["bank_holiday"]
+
+    meter_data = meter_data.drop(
+        ["tstp", "day_of_week", "day_type", "bank_holiday"], axis=1, errors="ignore"
+    )
+
+    # Calculate means for each half-hour and day type pair
+    meter_data_daytypes = (
+        meter_data.groupby(["weekend_or_bank_holiday", "time"])
+        .mean(numeric_only=True)
+        .dropna(axis=1)
+        .T
+    )
+
+    if normalise:
+        meter_data_daytypes_norm = meter_data_daytypes.div(
+            meter_data_daytypes.groupby(level=0, axis=1).transform("sum"), axis=0
+        ).dropna(axis=0)
```

Comment:
I included this just in case there were households that used no electricity and we ended up with 0/0 = NaN. Not sure there are any here but maybe worth keeping so this can be adapted for other datasets